New mountaineering module

AdequateUsername opened the pull request:
Adds some mountaineering-related gear to the game: skis, poles and crampons.

**Skis**
- crafted with iron ingots in the two side columns and iron boots in the centre of CC (advancement for crafting)
- give slowness on non-snowy blocks
- give speed on snowy blocks
- give increased speed when moving downhill (advancement for reaching maximum speed boost)
- create some particles when skiing

**Poles**
- crafted with a stick in top centre and bottom centre and tripwire hook in the centre of CC
- need to hold in mainhand and offhand for effects
- give jump boost 1 normally, and 2 when skiing uphill; this is on a cooldown
- reduce the vertical speed threshold for getting speed boosts when skiing
- reduce fall damage when used without skis

**Crampons**
- crafted with chainmail boots in top centre, and iron nuggets in centre row of CC
- give slowness effects (they shouldn't be easy to walk in)
- allow climbing of vertical walls made of stone, grass block, or dirt when both hands are empty
- you cannot 'climb' on a wall indefinitely; after a while you fall off
- the climbing mechanism is not perfect because (I think) of shulkers; _it is assumed that all climbers make mistakes_

**Remarks**
- ~~I can't see any changes needed to meet BPR's CC update, in that what I have looks similar to other modules. I might have missed something though~~ ~~I realise some changes are needed; going to look at this~~ CC should all be correct now
- Orbis compatibility would be good, but it looks like it is still being updated, so I'll leave this for now

Bloo-dev commented on gm4_mountaineering/data/gm4/advancements/mountaineering.json:
```diff
@@ -0,0 +1,43 @@
+      }
+    }
+  }
+}
```
Missing NL at EoF.

Bloo-dev commented on gm4_mountaineering/data/gm4_custom_crafters/tags/functions/check_recipes.json:
```diff
@@ -0,0 +1,5 @@
+    "values": [
+      "gm4_mountaineering:check_recipes"
+    ]
+  }
```
Missing NL at EoF.

Bloo-dev commented on gm4_mountaineering/data/gm4_mountaineering/functions/check_recipes.mcfunction:
```diff
@@ -0,0 +1,5 @@
+
+execute if score $crafted gm4_crafting matches 0 store success score $crafted gm4_crafting if score $slot_count gm4_crafting matches 4 if score $stack_size gm4_crafting matches ..1 if data storage gm4_custom_crafters:temp/crafter {Items:[{Slot:1b,id:"minecraft:iron_boots",Count:1b},{Slot:3b,id:"minecraft:heavy_weighted_pressure_plate",Count:1b},{Slot:4b,id:"minecraft:heavy_weighted_pressure_plate ...
+execute if score $crafted gm4_crafting matches 0 store success score $crafted gm4_crafting if score $slot_count gm4_crafting matches 3 if score $stack_size gm4_crafting matches ..1 if data storage gm4_custom_crafters:temp/crafter {Items:[{Slot:1b,id:"minecraft:tripwire_hook",Count:1b},{Slot:4b,id:"minecraft:iron_bars",Count:1b},{Slot:7b,id:"minecraft:pointed_dripstone",Count:1b}]} unless data sto ...
+execute if score $crafted gm4_crafting matches 0 store success score $crafted gm4_crafting if score $slot_count gm4_crafting matches 4 if score $stack_size gm4_crafting matches ..1 if data storage gm4_custom_crafters:temp/crafter {Items:[{Slot:1b,id:"minecraft:chainmail_boots",Count:1b},{Slot:3b,id:"minecraft:pointed_dripstone",Count:1b},{Slot:4b,id:"minecraft:pointed_dripstone",Count:1b},{Slot:5 ...
```
Missing NL at EoF.

Bloo-dev commented on gm4_mountaineering/data/gm4_mountaineering/functions/new_crampons.mcfunction:
```diff
@@ -0,0 +1,12 @@
+loot replace block ~ ~ ~ container.0 loot gm4_mountaineering:crafting/crampons
+
+# Remove temp-stored leather boots from stored data
+data remove storage gm4_mountaineering:temp crampons_item
```
Missing NL at EoF.

Bloo-dev commented on gm4_mountaineering/data/gm4_mountaineering/functions/new_skis.mcfunction:
```diff
@@ -0,0 +1,12 @@
+loot replace block ~ ~ ~ container.0 loot gm4_mountaineering:crafting/skis
+
+# Remove temp-stored iron boots from stored data
+data remove storage gm4_mountaineering:temp ski_item
```
Missing NL at EoF.

Bloo-dev commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/crampons_equipped.mcfunction:
```diff
@@ -0,0 +1,10 @@
+effect give @s jump_boost 2 0 true
+
+# Provide jump boost 2 when also holding poles
+effect give @s[predicate=gm4_mountaineering:holding_poles] jump_boost 2 1 true
```
Missing NL at EoF.

Bloo-dev commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/poles_equipped.mcfunction:
```diff
@@ -0,0 +1,6 @@
+# at @s
+# run from main
+
+execute at @a unless block ~ ~-0.2 ~ minecraft:air run effect give @s jump_boost 1 0 true
```
Missing NL at EoF.

Bloo-dev replied:
The `@a` here creates issues, you probably meant `@s`; regardless, do this location setting in `main`.
Creating a cooldown via the `unless block air` check is possible, however, only with high clock speeds, which nullifies the cooldown effect. If you want a cooldown, implement one using (e.g.) scores.

Bloo-dev commented on gm4_mountaineering/data/gm4_mountaineering/loot_tables/crafting/crampons.json:
```diff
@@ -0,0 +1,38 @@
+        ]
+      }
+    ]
+  }
```
Missing NL at EoF.

Bloo-dev commented on gm4_mountaineering/data/gm4_mountaineering/loot_tables/crafting/poles.json:
```diff
@@ -0,0 +1,29 @@
+        ]
+      }
+    ]
+  }
```
Missing NL at EoF.

Bloo-dev commented on gm4_mountaineering/data/gm4_mountaineering/loot_tables/crafting/skis.json:
```diff
@@ -0,0 +1,38 @@
+        ]
+      }
+    ]
+  }
```
Missing NL at EoF.

Bloo-dev commented on gm4_mountaineering/data/gm4_mountaineering/loot_tables/items/crampons.json:
```diff
@@ -0,0 +1,47 @@
+            ]
+        }
+    ]
+}
```
Missing NL at EoF.

Bloo-dev commented on gm4_mountaineering/data/gm4_mountaineering/loot_tables/items/poles.json:
```diff
@@ -0,0 +1,33 @@
+            ]
+        }
+    ]
+}
```
Missing NL at EoF.

Bloo-dev commented on gm4_mountaineering/data/gm4_mountaineering/loot_tables/items/skis.json:
```diff
@@ -0,0 +1,47 @@
+            ]
+        }
+    ]
+}
```
Missing NL at EoF.

Bloo-dev commented on gm4_mountaineering/data/gm4_mountaineering/predicates/holding_poles.json:
```diff
@@ -0,0 +1,31 @@
+      }
+    }
+  ]
+  
```
Missing NL at EoF.

Bloo-dev commented on gm4_mountaineering/data/gm4_mountaineering/predicates/wearing_crampons.json:
```diff
@@ -0,0 +1,15 @@
+      }
+    }
+  }
+  
```
Missing NL at EoF.

Bloo-dev commented on gm4_mountaineering/data/gm4_mountaineering/predicates/wearing_skis.json:
```diff
@@ -0,0 +1,14 @@
+        }
+      }
+    }
+}
```
Missing NL at EoF.

Bloo-dev commented on gm4_mountaineering/pack.mcmeta:
```diff
@@ -0,0 +1,33 @@
+        "Creator": [
+            "foodiebonus"
+        ],
+        "Updated By": [
+            "nobody"
+        ],
```
```suggestion
        "Updated By": [
        ],
```
Leave this array blank.

Bloo-dev commented on gm4_mountaineering/pack.mcmeta:
```diff
@@ -0,0 +1,33 @@
+        "Updated By": [
+            "nobody"
+        ],
+        "Icon Design": [
+            "nobody"
+        ]
```
```suggestion
        "Icon Design": [
        ]
```
Leave this array blank.

Bloo-dev commented on gm4_mountaineering/data/gm4_mountaineering/functions/main.mcfunction:
```diff
@@ -0,0 +1,8 @@
+# Run functions to provide relevant effects when a piece of mountaineering gear is worn
+
+execute as @a[gamemode=!spectator,gamemode=!adventure,predicate=gm4_mountaineering:wearing_skis] run function gm4_mountaineering:ski_effects/skis_equipped
+execute as @a[gamemode=!spectator,gamemode=!adventure,predicate=gm4_mountaineering:holding_poles] run function gm4_mountaineering:ski_effects/poles_equipped
+execute as @a[gamemode=!spectator,gamemode=!adventure,predicate=gm4_mountaineering:wearing_crampons] run function gm4_mountaineering:ski_effects/crampons_equipped
```
All of these should work adventure mode.
The functions should also be executed in the correct context here, consider adding `at @s` location setters where necessary and remove the context setters in the called functions.

Bloo-dev commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/skis_equipped.mcfunction:
```diff
@@ -0,0 +1,14 @@
+effect give @s[predicate=gm4_mountaineering:on_snowy_block] speed 2 0 true
+
+# Give extra speed if descending
+execute at @a if block ~ ~-1.3 ~ minecraft:air run effect give @s speed 2 1 true
+
+# Clear speed effects and apply slowness if standing on a non-snowy block
+execute at @a unless predicate gm4_mountaineering:on_snowy_block_or_air run effect clear @s minecraft:speed
+execute at @a unless predicate gm4_mountaineering:on_snowy_block_or_air run effect give @s slowness 3 0 true
```
Establish the location context in `main`, get rid of the `at @a` here.

Bloo-dev commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/skis_equipped.mcfunction:
```diff
@@ -0,0 +1,14 @@
+effect give @s[predicate=gm4_mountaineering:on_snowy_block] speed 2 0 true
+
+# Give extra speed if descending
+execute at @a if block ~ ~-1.3 ~ minecraft:air run effect give @s speed 2 1 true
```
The `@a` here creates issues, you probably meant `@s`; regardless, do this location setting in `main`.
The `if block air` check gives a lot of false positives when it comes to detecting falling, consider another method of checking for this -- think: what exactly do I want to detect? Players running down slopes?

Bloo-dev reviewed (changes requested):
First of all, thank you for your PR! We like the basic principles of this module and would love to see it be a part of Gamemode 4, especially with added Orbis compatibility to make these appear in loot chests!

Our in-game tests have shown some issues and necessary feature changes that are outlined below.
We have also found some more technical issues with the code, however, these may not need to be addressed individually, as some parts of this module should receive some refactoring to improve the end user experience. Though below you can still find a couple of out suggestions; bear in mind that this is not a comprehensive code review, though.

### Skies:
- the block checks need to be refined, the `stepping_on` condition sadly only checks the block below the player, neglecting, for example, snow layers entirely. This should be replaced with a smarter "snow detection system" which respects snow layers as a snowy surface, regardless of the block below the snow layer.
- the falling condition seems to be somewhat flawed (`air` 1.3 blocks below the player yields plenty of false positives). Use the scoreboard objective `minecraft.custom:minecraft.fall_one_cm` or advancement condition `minecraft:fall_from_height` instead, however, bear in mind that these only yield values once the player has landed making them less useful as it seems like what you want to track is players having a negative y-velocity (running downhill). The latter may be achieved by spawning markers and reading their y-coordinate to get a rough velocity approximation.
- instead of using potion effects, consider using attributes instead to apply speed (allows for more precise control, better stacking control)
- it would be nice if skis ramped up in speed when going downhill and if the skis emulated the slipperiness of skiing instead of the hard turns you can do when walking

### Poles & Crampons:
- Whereas the skis have their unique niche and have the interesting snow interaction the crampons & poles seem to be lacking a bit. We suggest spicing them up quite a bit to make them useful, here are our suggestions:
   -  Turn the crampons into boots that let you climb vertical walls by placing temporary Shulkers as temporary platforms.
   - Rework the poles -- tripwire hooks can't be used as the player can place them down. Instead, use an unplaceable item and increase the synergy with the skis (making you speed up quicker?). These should be early-game to give them the most use.

### Recipes
The current recipes seem a bit awkward and could use a bit of a re-vamp. Our points of concern are:
- iron pressure plates seem weird in a recipe
- dripstone seems weird in a recipe
Take some inspiration from our latest recipe re-vamp for custom crafters in which we fixed some of our old recipes.
This PR will also have to be updated to comply with @BluePsychoRanger's latest Custom Crafters update (#787 ). @BluePsychoRanger will give you a detailed review as well to add more info on this topic.

### Orbis Compatibility
Making these generate in Orbis Chests will advertise them to players, consider adding Orbis Compatibility. CC @kruthers who's currently updating Orbis.

AdequateUsername commented:
Firstly, thank you for the feedback. A quick update, in part to prove that I'm still working on this

- in general, I've given the skis better functionality: a better snow detection mechanism, a smarter way to detect falling (calculate y-velocity) and add speed (now with attributes, not potion effects) depending on v_y.
- re: emulating slipperiness, that was something I wanted to implement from the start but can't find a nice way of doing it. The way I envision it is that it's as if the player is on an ice block. I don't know if it's possible to set the block beneath the player to ice with the texture of whatever block is actually there, but that sounds rather messy.
- adjusted recipes, and changed the poles to be sticks (maybe too 'boring'?)

The adding-speed functionality generally follows the mechanism used in speed paths, but it feels a bit clunky/I don't know if it puts an unnecessary burden lag-wise. A way to increase the speed attribute as a function of the player's y-velocity would probably be the gold standard, but the current method didn't seem horrific so I went with it.

Further updates will come for poles/crampons to make them more whole, bearing in mind Bloo's feedback. And adjusting to conform to BPR's CC update + Orbis compatibility (but I need to learn how to do that first.. :) I can't guarantee any timeframe right now; sorry for that, but I haven't forgotten this.

TheThanathor commented:
Since you're still working on this I won't do a full in-depth review. On the functionality side I've tested the skis and they feel great to use now!

On the technical side you should attempt to limit the amount of `execute as @a[gamemode=!spectator] at @s`.
For example, after calling `gm4_mountaineering:ski_effects/check_snowy_block` you should put the `gm4_mountaineering:get_velocity` call in that function directly instead of calling it from `main` later.

Directly calling a function will keep any `as` or `at` context, so you only have to check for any additional conditions. This goes for pretty much all of the ski related function calls in `main`, where you can call from the checking function directly instead of adding a tag to the player then checking for players with that tag.

AdequateUsername commented:
Understood about pulling all of those checks out of main. I'll write the crampons section of the module in that way, then come back to fix the skis at the end (a decent chunk of reformatting is needed anyway). Thanks!

AdequateUsername commented:
This should be more or less ready for submission; I've updated the main post with full details of the module.
As always, thank you for the help so far

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/climbing_effects/damage_shulkers.mcfunction:
```diff
@@ -0,0 +1,10 @@
+scoreboard players remove @e[type=minecraft:shulker, tag=gm4_mountaineering_climb_assist] gm4_mountaineering_climbing_shulker_countdown 1
+tag @e[type=minecraft:shulker, tag=gm4_mountaineering_climb_assist, scores={gm4_mountaineering_climbing_shulker_countdown=..0}] add gm4_mountaineering_kill_shulker
+tag @e[type=minecraft:shulker, tag=gm4_mountaineering_climb_assist, distance=2.5..] add gm4_mountaineering_kill_shulker
+execute run function gm4_mountaineering:climbing_effects/kill_shulkers
```
I'm going to leave the comment here, but this is for all of the other files as well, you don't need the `execute run` part unless you're changing the `at` or `as` first
```suggestion
function gm4_mountaineering:climbing_effects/kill_shulkers
```

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/climbing_effects/kill_shulkers.mcfunction:
```diff
@@ -0,0 +1,7 @@
+# at @s
+# run from climbing_effects/damage_shulkers and climbing_effects/clear_climbing_effects
+
+teleport @e[type=minecraft:shulker, tag=gm4_mountaineering_kill_shulker] ~ ~-400 ~
+kill @e[type=minecraft:shulker, tag=gm4_mountaineering_kill_shulker]
```
This function should execute as the shulker that needs to be killed, so you don't need the `gm4_mountaineering_kill_shulker` tag and you can use `@s` instead.

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/get_velocity.mcfunction:
```diff
@@ -0,0 +1,14 @@
+scoreboard players operation @s gm4_mountaineering_vy -= @s gm4_mountaineering_y2
+
+# Update y2 for use on next run of this function
+scoreboard players operation @s gm4_mountaineering_y2 = @s gm4_mountaineering_y1
```
missing newline

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/main.mcfunction:
```diff
@@ -0,0 +1,13 @@
+# Run ski-related functions
+execute as @a[gamemode=!spectator] at @s run function gm4_mountaineering:ski_effects/skis_equipped
+
+# Adds jump boost effects when holding poles
+execute as @a[gamemode=!spectator,predicate=gm4_mountaineering:holding_poles] at @s run function gm4_mountaineering:ski_effects/poles_jump_boost
+
+# Adds vertical climbing abilities when wearing crampons and not holding anything
+execute as @a[gamemode=!spectator,predicate=gm4_mountaineering:wearing_crampons] at @s run function gm4_mountaineering:climbing_effects/crampons_equipped
+
+# Clear any crampon effects when crampons are removed
+execute as @a[gamemode=!spectator,predicate=!gm4_mountaineering:wearing_crampons] at @s run function gm4_mountaineering:climbing_effects/clear_crampon_effects
```
It would be better if you used a `player_submain` function instead, so you can reduce the amount of `@a at @s` you have to use.

TheThanathor replied:
This is run for every player not wearing Crampons all the time, it should only run once if you stop wearing them (you can do this by adding a tag when they start wearing Crampons and checking only for players with that tag)

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/main.mcfunction:
```diff
@@ -0,0 +1,13 @@
+execute as @a[gamemode=!spectator,predicate=gm4_mountaineering:holding_poles] at @s run function gm4_mountaineering:ski_effects/poles_jump_boost
+
+# Adds vertical climbing abilities when wearing crampons and not holding anything
+execute as @a[gamemode=!spectator,predicate=gm4_mountaineering:wearing_crampons] at @s run function gm4_mountaineering:climbing_effects/crampons_equipped
```
This function runs every time for players wearing Crampons, but not everything in the called function needs to run every time. It's better to call a separate "initiate Crampons" function first that adds a tag to the player and does things that only need to happen once like adding the speed increase Attribute.
Anything that does need to be on a clock can then just run on players with that tag.

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/check_snowy_block.mcfunction:
```diff
@@ -0,0 +1,15 @@
+execute unless block ~ ~-0.1 ~ minecraft:air run tag @s remove gm4_mountaineering_skiing_off_snow
+
+# Add skiing tag if on snow layer or any other snowy block (contained within the predicate) with skis
+execute if block ~ ~ ~ minecraft:snow if predicate gm4_mountaineering:wearing_skis run tag @s add gm4_mountaineering_skiing
+execute if predicate gm4_mountaineering:on_snowy_block if predicate gm4_mountaineering:wearing_skis run tag @s add gm4_mountaineering_skiing
```
You should be able to combine these two by adding a check for the snow block in your `on_snowy_block` predicate

AdequateUsername replied:
If I recall, the reason I had to split this up is because the predicate couldn't detect `snow`. (It detects `snow block` fine, but for some reason `snow`, referring to just the single layer, had to be called separately as done in this function)

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/check_snowy_block.mcfunction:
```diff
@@ -0,0 +1,15 @@
+# clear existing tags, check if the player is on a snowy block and wearing skis, add the skiing tag back if yes, and add ski_off_snow tag if not on snowy block
+# @s = player
+# at @s
+# run from main
```
this function is called from `gm4_mountaineering:ski_effects/skis_equipped`

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/main.mcfunction:
```diff
@@ -0,0 +1,13 @@
+# Run ski-related functions
+execute as @a[gamemode=!spectator] at @s run function gm4_mountaineering:ski_effects/skis_equipped
```
This should first check the `gm4_mountaineering:wearing_skis` predicate
```suggestion
execute as @a[gamemode=!spectator,predicate=gm4_mountaineering:wearing_skis] at @s run function gm4_mountaineering:ski_effects/skis_equipped
```

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/check_snowy_block.mcfunction:
```diff
@@ -0,0 +1,15 @@
+execute if predicate gm4_mountaineering:on_snowy_block if predicate gm4_mountaineering:wearing_skis run tag @s add gm4_mountaineering_skiing
+
+# Add a tag for skiing off a snowy block if wearing skis on a non-snowy block
+execute unless block ~ ~ ~ minecraft:snow unless predicate gm4_mountaineering:on_snowy_block_or_air if predicate gm4_mountaineering:wearing_skis run tag @s add gm4_mountaineering_skiing_off_snow
```
All of the `gm4_mountaineering:wearing_skis` predicate checks should not be needed, as the function that this is called from already only runs for players wearing skis

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/downhill_ski_speed/add_downhill_ski_speed.mcfunction:
```diff
@@ -0,0 +1,8 @@
+
+tag @s add gm4_mountaineering_skiing_downhill
+scoreboard players set @s gm4_mountaineering_ski_downhill 4
+attribute @s minecraft:generic.movement_speed modifier add 42a046eb-1425-4efe-831b-f891d8ede18f gm4_mountaineering_downhill_speed_boost 0.150 multiply_base
```
I won't leave this comment on other attribute commands, but you should look through the rest and fix them as well
```suggestion
attribute @s minecraft:generic.movement_speed modifier add 42a046eb-1425-4efe-831b-f891d8ede18f "gm4_mountaineering_downhill_speed_boost" 0.150 multiply_base
```

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/downhill_ski_speed/check_downhill_ski_speed.mcfunction:
```diff
@@ -0,0 +1,11 @@
+# Checks that the player still travels downhill (>=2 blocks)
+# @s = player with skiing tag and having fallen >=2 blocks
+# at @s
+# run from main
```
You have a few of these headers with the wrong `run from`, and I found some as well that said they ran for "a player wearing skis" but actually ran for all players, make sure to carefully check these as a wrong header makes it easy to get confused on why things are not working as you think they should.

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/poles_jump_boost.mcfunction:
```diff
@@ -0,0 +1,15 @@
+# run from main
+
+# Add jump boost effect (1 regularly, 2 if having an upwards velocity)
+execute if score @s gm4_mountaineering_poles_jump_boost matches ..0 if score @s gm4_mountaineering_vy matches 2.. run effect give @s jump_boost 1 1 true
+execute if score @s gm4_mountaineering_poles_jump_boost matches ..0 if score @s gm4_mountaineering_vy matches ..0 run effect give @s jump_boost 1 0 true
+
+# Reduce jump cooldown
+execute if score @s gm4_mountaineering_poles_jump_boost matches 1.. run scoreboard players remove @s gm4_mountaineering_poles_jump_boost 1
+
+# If jump, reset cooldown and jump counter
+execute if entity @s[scores={gm4_mountaineering_jump_counter=1..}] run scoreboard players set @s gm4_mountaineering_poles_jump_boost 3
+execute if entity @s[scores={gm4_mountaineering_jump_counter=1..}] run scoreboard players reset @s gm4_mountaineering_jump_counter
```
I'm not sure about the whole cooldown aspect of this, it felt kind of awkward to use in-game, especially getting jump boost II was rare as to climb and get the boost you had to jump which put it on cooldown.

AdequateUsername replied:
The reason for the cooldown is because, even with poles, I don't see them as a tool for making many jumps with - only occasionally. Hence the desire to prevent them being used as such.

The jump boost II use case was made more with the intention of a ski-jumping-ramp (e.g. slabs or snow layers) to allow you to get a large upward velocity, then 'launch' yourself off that. I can get rid of it if it's annoying, but at least the reasoning is clear.

(I've also added some extra poles functionality, which I'll summarise in the main post later)

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/poles_jump_boost.mcfunction:
```diff
@@ -0,0 +1,15 @@
+execute if score @s gm4_mountaineering_poles_jump_boost matches 1.. run scoreboard players remove @s gm4_mountaineering_poles_jump_boost 1
+
+# If jump, reset cooldown and jump counter
+execute if entity @s[scores={gm4_mountaineering_jump_counter=1..}] run scoreboard players set @s gm4_mountaineering_poles_jump_boost 3
+execute if entity @s[scores={gm4_mountaineering_jump_counter=1..}] run scoreboard players reset @s gm4_mountaineering_jump_counter
```
No need to check scores here unless you don't want to reset players with a score below 0
```suggestion
execute if entity @s[scores={gm4_mountaineering_jump_counter=1..}] run scoreboard players set @s gm4_mountaineering_poles_jump_boost 3
scoreboard players reset @s gm4_mountaineering_jump_counter
```

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/skis_equipped.mcfunction:
```diff
@@ -0,0 +1,28 @@
+execute run function gm4_mountaineering:get_velocity
+
+# Adds speed effects for skiing on a flat surface
+execute if entity @s[tag=gm4_mountaineering_skiing] run function gm4_mountaineering:ski_effects/flat_ski_speed/add_flat_ski_speed
+execute if entity @s[tag=!gm4_mountaineering_skiing] run function gm4_mountaineering:ski_effects/flat_ski_speed/remove_flat_ski_speed
```
You do this quite a few more times, but I'll leave the comment here. These functions will keep running every clock cycle. You should add a tag to the player in `add_flat_ski_speed`, I'll use `gm4_mountaineering_ski_speed` as an example, and remove it in `remove_flat_ski_speed`
```suggestion
execute if entity @s[tag=gm4_mountaineering_skiing,tag=!gm4_mountaineering_ski_speed] run function gm4_mountaineering:ski_effects/flat_ski_speed/add_flat_ski_speed
execute if entity @s[tag=!gm4_mountaineering_skiing,tag=gm4_mountaineering_ski_speed] run function gm4_mountaineering:ski_effects/flat_ski_speed/remove_flat_ski_speed
```

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/steep_downhill_ski_speed/check_steep_downhill_ski_speed.mcfunction:
```diff
@@ -0,0 +1,13 @@
+execute if score @s[tag=gm4_mountaineering_skiing,predicate=gm4_mountaineering:holding_poles] gm4_mountaineering_vy matches ..-7 run scoreboard players set @s gm4_mountaineering_ski_steep_downhill 4
+
+# Remove speed if no longer travelling downhill or removed skis
+execute at @s[scores={gm4_mountaineering_ski_steep_downhill=..0}] run function gm4_mountaineering:ski_effects/steep_downhill_ski_speed/remove_steep_downhill_ski_speed
+execute at @s[tag=!gm4_mountaineering_skiing] run function gm4_mountaineering:ski_effects/steep_downhill_ski_speed/remove_steep_downhill_ski_speed
```
This is a thing in some other files as well, but if you're already executing `at @s` (which this function is according to your header) you only need to use another `at @s` if you teleport the player. Instead, us `if entity @s` here.

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/predicates/hands_empty.json:
```diff
@@ -0,0 +1,19 @@
+      }
+    }
+  }
+  
```
missing newline

TheThanathor reviewed (changes requested):
Module is looking greatly improved since last time! Really like the direction the items went, they feel much nicer to use, and the code is looking a lot better as well.
There's still a bit to do, the most important thing to reduce is when you call functions that are not needed.

**Skis**
These feel great to use now, it might be worth looking into some particle or even sound effects to spruce the up a little bit, but that might also quickly before too much. Experiment a little and see what you like.

**Crampons**
Love the idea of these now, and they're close to being great, but I think they need a faster clock to work properly. I would also like for you to be able to sit on a wall for a while longer, as taking in the view is a large part of the fun.

**Poles**
The Poles seem a little bit weak right now, I'd suggest removing the cooldown from them and maybe giving them more interactions, they could give a small speed bonus in general for instance to make them useful for general exploring too (though that might diminish the "mountaineering" aspect of them)

I've left quite a few comments and I'm not exactly a master of this stuff myself so if you are confused about anything feel free to ask me / others on the discord.

AdequateUsername commented on gm4_mountaineering/data/gm4_mountaineering/functions/climbing_effects/damage_shulkers.mcfunction:
```diff
@@ -0,0 +1,10 @@
+# decrease the score of each climbing shulker
```
I've simplified the function a lot:
- damage_shulkers gets called independently of crampons_equipped, so all climbing shulkers are affected regardless of what the player is doing
- damage_shulkers increases each shulker's score by 1
- when reaching 40, the shulker is killed
- damage_shulkers now executed at climbing shulkers

AdequateUsername commented:
Thank you for the detailed feedback! I've adjusted the code for most of the issues; for those that I haven't I've left a brief explanation to justify that. The crafting system isn't working quite as I thought it was - I'll spend some time soon to try and work it out, but I think probably I'll have to ask BPR.

Aside from the adjustments from that feedback, poles now reduce fall damage (I don't see them really as a way to get more speed, but they do help with general stability when walking - fall damage ties in nicely), and there are some particles when skiing. I also added an advancement for reaching maximum speed boost, but I'm not good with names so it's somewhat dull.

AdequateUsername commented:
I think the custom crafting is now all working on the latest version.
That should be everything for this feedback round; I've re-requested a review, but there's no hurry from my side.

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/pole_effects/poles_jump_boost.mcfunction:
```diff
@@ -0,0 +1,15 @@
+# run from poles_equipped
+
+# Add jump boost effect (1 regularly, 2 if having an upwards velocity)
+execute if score @s gm4_mountaineering_poles_jump_boost matches ..0 if score @s gm4_mountaineering_vy matches 2.. run effect give @s jump_boost 1 1 true
+execute if score @s gm4_mountaineering_poles_jump_boost matches ..0 if score @s gm4_mountaineering_vy matches ..0 run effect give @s jump_boost 1 0 true
```
jump boost level 1 feels rather inconsequential, I would recommend changing the base level to 2, then use 3-4 for the increased one. You might also want to reduce the `gm4_mountaineering_vy` threshold a bit as it felt really difficult to reach the higher-level jump boost

AdequateUsername replied:
Reducing the threshold would be a bit difficult to do without reworking the whole vy system (since it only works with integers). So I've done this instead:
- if you wear skis, you get jump boost 1 (same as current system); jumping while wearing skis is a bit difficult
- if you don't wear skis, you get jump boost 2
- regardless of whether you wear skis or not, you get jump boost 3 when moving uphill; this much larger jump boost (I think) should compensate for the fact that you don't get it all the time

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/check_snowy_block.mcfunction:
```diff
@@ -0,0 +1,15 @@
+# run from skis_equipped
+
+# Remove tags unless the player is in the air
+execute unless block ~ ~-0.1 ~ minecraft:air run tag @s remove gm4_mountaineering_skiing
+execute unless block ~ ~-0.1 ~ minecraft:air run tag @s remove gm4_mountaineering_skiing_off_snow
```
this can call the `gm4_mountaineering:ski_effects/ski_off_snow_slowness/remove_off_snow_slowness`, and where you add the tag in this function you can call `gm4_mountaineering:ski_effects/ski_off_snow_slowness/add_off_snow_slowness`, which would remove the need for those function tags in `skis_equipped`. Make sure you move the tag checks here so it doesn't run when it doesn't need to

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/skis_equipped.mcfunction:
```diff
@@ -0,0 +1,32 @@
+execute if entity @s[tag=gm4_mountaineering_skiing_downhill] run function gm4_mountaineering:ski_effects/downhill_ski_speed/check_downhill_ski_speed
+
+# Adds greater increased speed effects for skiing on a steep downhill surface, with a lower vertical velocity threshold if poles are used
+execute if entity @s[tag=gm4_mountaineering_skiing,tag=!gm4_mountaineering_skiing_steep_downhill,scores={gm4_mountaineering_vy=..-7},predicate=!gm4_mountaineering:holding_poles] run function gm4_mountaineering:ski_effects/steep_downhill_ski_speed/add_steep_downhill_ski_speed
```
the `predicate=!gm4_mountaineering:holding_poles` should not be necessary here

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/steep_downhill_ski_speed/add_steep_downhill_ski_speed.mcfunction:
```diff
@@ -0,0 +1,11 @@
+attribute @s minecraft:generic.movement_speed modifier add 8aed075d-afeb-4943-af9d-6b7e1d31e0de "gm4_mountaineering_steep_downhill_speed_boost" 0.150 multiply_base
+
+# Give an advancement if not yet achieved
+execute if entity @s[advancements={gm4:ski=true}] run advancement grant @s only gm4:ski_fast
```
```suggestion
advancement grant @s only gm4:ski_fast
```
Generally advancements should not care about the previous advancement having been obtained

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/loot_tables/crafting/crampons.json:
```diff
@@ -0,0 +1,30 @@
+      }
+    ]
+  }
+  
```
missing newline

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/tags/blocks/climbable_blocks.json:
```diff
@@ -0,0 +1,7 @@
+    "minecraft:dirt",
+    "minecraft:grass_block"
+    ]
+   }
```
missing newline

Stone variants should also be included here, you can use `#minecraft:base_stone_overworld` to cover them all. I feel like the ice variants should also be included, maybe things like ores as well in case they're in the wall. Probably some more that I missed

TheThanathor commented on gm4_mountaineering/pack.mcmeta:
```diff
@@ -0,0 +1,35 @@
+        "lib_forceload"
+    ],
+    "site_description": "A mezze of mountaineering means!",
+    "video_link": "YOUTUBE_VIDEO_URL",
+    "wiki_link": "WIKI_URL",
```
```suggestion
    "wiki_link": "https://wiki.gm4.co/wiki/Mounteneering",
```

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/skis_equipped.mcfunction:
```diff
@@ -0,0 +1,32 @@
+function gm4_mountaineering:ski_effects/check_snowy_block
+
+# Adds particles at the player's feet when skiing on a snowy bock and without air below
+execute if entity @s[tag=gm4_mountaineering_skiing] unless block ~ ~-2 ~ minecraft:air run particle minecraft:cloud ~ ~ ~ 0.1 0.1 0.1 0.001 50 normal
```
particles feel way too much, try reducing count to like 2-3. Maybe a different particle altogether. It should feel more like the running particles

TheThanathor reviewed (commented):
Module is looking great! Getting pretty close to being able to add this to the server. Left a few more mostly minor comments, after that you'll probably want a review from BPR or someone else that knows about the new Custom Crafters to check if that's all up to par.

AdequateUsername commented on gm4_mountaineering/data/gm4_mountaineering/functions/climbing_main.mcfunction:
```diff
@@ -0,0 +1,9 @@
+execute as @a[gamemode=!spectator,tag=gm4_mountaineering_using_crampons] at @s run function gm4_mountaineering:climbing_effects/crampons_equipped
+
+# Apply damage to any climbing shulkers
+execute as @e[type=minecraft:shulker,tag=gm4_mountaineering_climb_assist] at @s run function gm4_mountaineering:climbing_effects/damage_shulkers
```
Assuming this refers to the damaging shulker part - if I moved it to the main clock the shulkers would be killed off 4x slower, which I wanted to avoid

TheThanathor replied:
You can reduce the score before the kill by roughly 1/4th to keep the same duration

AdequateUsername replied:
Oh, of course - for some reason I was thinking the damaging system was something different :/

AdequateUsername commented on gm4_mountaineering/data/gm4_mountaineering/functions/pole_effects/poles_equipped.mcfunction:
```diff
@@ -0,0 +1,10 @@
+function gm4_mountaineering:pole_effects/poles_jump_boost
+
+execute if entity @s[nbt={OnGround:0b}] if block ~ ~-2 ~ minecraft:air run effect give @s resistance 1 1
+execute if entity @s[nbt={OnGround:1b}] run effect clear @s resistance
```
I've moved the resistance-adding to the faster clock (which is otherwise used for the climbing stuff), so it's quick enough that NBT tags aren't needed

AdequateUsername commented:
Okay, everything should be updated to meet the latest feedback - thank you (again) for giving it. I'll ask BPR for a crafting-focused review to check that's all fine.e

BluePsychoRanger commented on gm4_mountaineering/data/gm4_mountaineering/functions/climbing_effects/damage_shulkers.mcfunction:
```diff
@@ -0,0 +1,7 @@
+# run from player_main
+
+scoreboard players add @s gm4_mountaineering_climbing_shulker_countdown 1
+execute as @s[scores={gm4_mountaineering_climbing_shulker_countdown=6..}] run function gm4_mountaineering:climbing_effects/kill_shulkers
```
Functions have a bit of an overhead, so it'd probably be better to just run the kill and tp commands here.
```suggestion
teleport @s[scores={gm4_mountaineering_climbing_shulker_countdown=6..}] ~ -10000 ~
kill @s[scores={gm4_mountaineering_climbing_shulker_countdown=6..}]
```

BluePsychoRanger commented on gm4_mountaineering/data/gm4_mountaineering/functions/climbing_effects/kill_shulkers.mcfunction:
```diff
@@ -0,0 +1,7 @@
+# kill climbing shulkers
+# @s = shulker with tag gm4_mountaineering_climb_assist and needing to be killed
```
this is quite a small function, so it's probably just better to put the commands directly in `damage_shulkers`

BluePsychoRanger commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/check_snowy_block.mcfunction:
```diff
@@ -0,0 +1,16 @@
+
+# Add skiing tag if on snow layer or any other snowy block (contained within the predicate) with skis
+execute if block ~ ~ ~ minecraft:snow run tag @s add gm4_mountaineering_skiing
+execute if predicate gm4_mountaineering:on_snowy_block run tag @s add gm4_mountaineering_skiing
```
```suggestion
tag @s[predicate=gm4_mountaineering:on_snowy_block] add gm4_mountaineering_skiing
```

BluePsychoRanger commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/steep_downhill_ski_speed/check_steep_downhill_ski_speed.mcfunction:
```diff
@@ -0,0 +1,13 @@
+# at @s
+# run from skis_equipped
+
+execute if entity @s[scores={gm4_mountaineering_ski_steep_downhill=1..},predicate=!gm4_mountaineering:holding_poles] unless score @s gm4_mountaineering_vy matches ..-7 run scoreboard players remove @s gm4_mountaineering_ski_steep_downhill 1
+execute if entity @s[scores={gm4_mountaineering_ski_steep_downhill=1..},predicate=gm4_mountaineering:holding_poles] unless score @s gm4_mountaineering_vy matches ..-5 run scoreboard players remove @s gm4_mountaineering_ski_steep_downhill 1
```
Why do you do score checks two different ways? the `unless score ...` subcommand can be combined with the `if entity ...` subcommand

BluePsychoRanger commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/steep_downhill_ski_speed/check_steep_downhill_ski_speed.mcfunction:
```diff
@@ -0,0 +1,13 @@
+
+execute if entity @s[scores={gm4_mountaineering_ski_steep_downhill=1..},predicate=!gm4_mountaineering:holding_poles] unless score @s gm4_mountaineering_vy matches ..-7 run scoreboard players remove @s gm4_mountaineering_ski_steep_downhill 1
+execute if entity @s[scores={gm4_mountaineering_ski_steep_downhill=1..},predicate=gm4_mountaineering:holding_poles] unless score @s gm4_mountaineering_vy matches ..-5 run scoreboard players remove @s gm4_mountaineering_ski_steep_downhill 1
+execute if score @s[tag=gm4_mountaineering_skiing] gm4_mountaineering_vy matches ..-7 run scoreboard players set @s gm4_mountaineering_ski_steep_downhill 4
+execute if score @s[tag=gm4_mountaineering_skiing,predicate=gm4_mountaineering:holding_poles] gm4_mountaineering_vy matches ..-7 run scoreboard players set @s gm4_mountaineering_ski_steep_downhill 4
```
```suggestion
scoreboard players set @s[tag=gm4_mountaineering_skiing,scores={gm4_mountaineering_vy=..-7}]  gm4_mountaineering_ski_steep_downhill 4
scoreboard players set @s[tag=gm4_mountaineering_skiing,predicate=gm4_mountaineering:holding_poles,scores={gm4_mountaineering_vy=..-7}] gm4_mountaineering_ski_steep_downhill 4
```

BluePsychoRanger replied:
The first two commands can be rewritten like this as well

BluePsychoRanger commented on gm4_mountaineering/data/gm4_mountaineering/functions/ski_effects/downhill_ski_speed/check_downhill_ski_speed.mcfunction:
```diff
@@ -0,0 +1,11 @@
+# at @s
+# run from skis_equipped
+
+execute if entity @s[scores={gm4_mountaineering_ski_downhill=1..}] unless score @s gm4_mountaineering_vy matches ..-2 run scoreboard players remove @s gm4_mountaineering_ski_downhill 1
+execute if score @s[tag=gm4_mountaineering_skiing] gm4_mountaineering_vy matches ..-2 run scoreboard players set @s gm4_mountaineering_ski_downhill 4
```
```suggestion
scoreboard players remove @s[scores={gm4_mountaineering_ski_downhill=1..,gm4_mountaineering_vy=-1..}] gm4_mountaineering_ski_downhill 1
scoreboard players set @s[tag=gm4_mountaineering_skiing,scores={gm4_mountaineering_vy=..-2}] gm4_mountaineering_ski_downhill 4
```
It's best to use as little execute subcommands as possible

BluePsychoRanger commented on gm4_mountaineering/data/gm4_mountaineering/functions/pole_effects/poles_equipped.mcfunction:
```diff
@@ -0,0 +1,16 @@
+# run from player_main
+
+# Add jump boost effect (1 with skis, 2 without skis, 3 if having an upwards velocity)
+execute if score @s gm4_mountaineering_poles_jump_boost matches ..0 if score @s gm4_mountaineering_vy matches 2.. run effect give @s jump_boost 1 2 true
```
```suggestion
effect give @s[scores={gm4_mountaineering_poles_jump_boost=..0, gm4_mountaineering_vy=2..}] jump_boost 1 2 true
```

BluePsychoRanger replied:
It's best to use as few execute subcommands as possible. I'll let you figure out how to fix the rest in this function

BluePsychoRanger commented on gm4_mountaineering/data/gm4_mountaineering/functions/pole_effects/poles_equipped.mcfunction:
```diff
@@ -0,0 +1,16 @@
+
+# If jump, reset cooldown and jump counter
+execute if entity @s[scores={gm4_mountaineering_jump_counter=1..}] run scoreboard players set @s gm4_mountaineering_poles_jump_boost 3
+scoreboard players reset @s gm4_mountaineering_jump_counter
```
This should be called for `@a` at the end of `main.mcfunction` so it just clears all players instead of running the same command for each player

BluePsychoRanger reviewed (changes requested):
Looks good functionality-wise, but I left a bunch of optimization comments. Don't be discouraged cuz most of the things I pointed out are almost negligible and only matter when there are a lot of functions being run at once, which is what usually happens when GM4 datapacks are used.

AdequateUsername commented on .vscode/settings.json:
```diff
@@ -3,5 +3,11 @@
+	],
+	"githubPullRequests.ignoredPullRequestBranches": [
+		"master"
+	]
```
That's really weird, I don't know why that changed. I reverted it back.

AdequateUsername commented on gm4_mountaineering/data/gm4_mountaineering/functions/player_main.mcfunction:
```diff
@@ -0,0 +1,22 @@
+# run from main
+
+# Run ski-related functions
+execute at @s if entity @s[predicate=gm4_mountaineering:wearing_skis] run function gm4_mountaineering:ski_effects/skis_equipped
```
I've gone for `at @s[predicate=gm4_mountaineering:wearing_skis]` now, so there aren't any options for putting `at @s` in any particular order.

I take it this is the best way? Or did you mean to put `at @s` and then the `if entity` subcommand? (I assumed not the latter, since we want to keep subcommands down)

AdequateUsername commented on gm4_mountaineering/data/gm4_mountaineering/functions/player_main.mcfunction:
```diff
@@ -0,0 +1,22 @@
+execute at @s if entity @s[predicate=!gm4_mountaineering:wearing_crampons,tag=gm4_mountaineering_using_crampons] run tag @s remove gm4_mountaineering_using_crampons
+
+# Apply damage to any climbing shulkers
+execute as @e[type=minecraft:shulker,tag=gm4_mountaineering_climb_assist] at @s run function gm4_mountaineering:climbing_effects/damage_shulkers
```
I didn't think it really fit anywhere else (since the climbing functions are generally all run from the faster clock, but this doesn't need to be so isn't)
re: distance check, I'm not too sure how to implement this. I tried putting in `[distance=2..]` inside the `@e` call and following `@s` (which I thought would make sense) but that doesn't work. I'm also not sure how I'd make it unique to the player that made it - for example, shulkers and being used by one player shouldn't be getting damaged because another player is on the same server (and greater than 2 blocks away). So for now, I just have it that all shulkers get damaged and killed off regardless of how a player uses them (justified by saying climbers can't hold on to a wall forever).

BluePsychoRanger replied:
it should go in `main`. Since it's here, the function is being called per player meaning the more players there are on the server the faster the shulkers die... if there are 6 or more players online shulkers die immediately on the next clock cycle.

BluePsychoRanger replied:
If this is moved to `main` every command uses `at @s`, so the `at @s` should be moved to `main` when calling `player_main` and these commands should just use `if entity`

AdequateUsername commented on gm4_mountaineering/data/gm4_mountaineering/loot_tables/items/poles.json:
```diff
@@ -0,0 +1,33 @@
+                                "translate": "%1$s%3427655$s",
+                                "with": [
+                                    "Poles",
+                                    {
```
If that would make their use more obvious, I'm happy to change that - but their use case isn't limited only to skiing, they can be used for walking also (as they help reduce some fall damage).

BluePsychoRanger replied:
I think it'd be best to use "Ski Poles." IRL a trekking pole is really just any sturdy stick which is why ski poles can also be used as trekking poles.

AdequateUsername replied:
Sounds good; change made

AdequateUsername commented on gm4_mountaineering/pack.mcmeta:
```diff
@@ -0,0 +1,34 @@
+    ],
+    "credits": {
+        "Creator": [
+            "foodiebonus"
```
Updated - but I don't have links, so didn't put that in.

AdequateUsername commented on gm4_mountaineering/data/gm4_mountaineering/functions/climbing_effects/add_climb_effects.mcfunction:
```diff
@@ -0,0 +1,14 @@
+effect give @s jump_boost 1 1 true 
+
+# summon shulker in front and below player, facing towards them
+execute if entity @s[y_rotation=135..180] positioned ~ ~-1.5 ~-1 align xyz unless entity @e[type=shulker, tag=gm4_mountaineering_climb_assist, dx=0, dz=0, dy=0] unless block ~0.5 ~ ~0.5 minecraft:air run summon minecraft:shulker ~0.5 ~ ~0.5 {Color:7,Peek:20,AttachFace:2,Invulnerable:1b,NoAI:1b,NoGravity:1b,PersistenceRequired:1b,Silent:1b,Tags:[gm4_mountaineering_climb_assist],DeathLootTable:"min ...
```
There is some visual glitch, yes, where the block goes dark for a tick when the shulker is first summoned. I've tried the teleporting solution, by summoning them 1000 blocks above the intended position, then running:

`execute as @e[type=shulker, tag=gm4_mountaineering_climb_assist, y=1000, dy=500] at @s run teleport @s ~ ~-1000 ~`
(dy=500 so the whole world height is included)
_but_ I think I may have misunderstood something with the context setters? The shulker gets teleported, but the player always seems to get lodged inside the shulker and can't move. Perhaps there's some delay so the player falls far enough that the shulker falls inside it? If I made an obvious mistake here, I'd be grateful if you could point it out, but if it's more complex then I'll keep thinking myself..

AdequateUsername replied:
// solved now, by adding another function to reset peek, which is scheduled 1t after teleporting

BluePsychoRanger commented on gm4_mountaineering/data/gm4_mountaineering/functions/climbing_effects/add_climb_effects.mcfunction:
```diff
@@ -1,14 +1,17 @@
+execute if entity @s[y_rotation=45..135] positioned ~-1 ~-1.5 ~ align xyz unless entity @e[type=shulker, tag=gm4_mountaineering_climb_assist, dx=0, dz=0, dy=0] unless block ~0.5 ~ ~0.5 minecraft:air run summon minecraft:shulker ~0.5 ~1000 ~0.5 {Color:7,Peek:20,AttachFace:4,Invulnerable:1b,NoAI:1b,NoGravity:1b,PersistenceRequired:1b,Silent:1b,Tags:[gm4_mountaineering_climb_assist],DeathLootTable:" ...
+
+# teleport shulker down 700 blocks to bring them into position
+execute as @e[type=shulker, tag=gm4_mountaineering_climb_assist, y=1000, dy=500] at @s run teleport @s ~ ~-1000 ~
```
```suggestion
execute positioned ~ ~1000 ~ as @e[type=shulker, tag=gm4_mountaineering_climb_assist] at @s run teleport @s ~ ~-1000 ~
```

AdequateUsername replied:
This still leads to some kind of strange behaviour - this command alone, as far as I can tell, has the shulkers continuously teleporting down (so they go past the player again). If I put in some kind of check (eg predicate for only entities above y=500) then they teleport to the same place, but don't seem to open properly and running into them in any way causes the player to get stuck.

BluePsychoRanger commented on gm4_mountaineering/data/gm4_mountaineering/functions/player_fast_main.mcfunction:
```diff
@@ -0,0 +1,8 @@
+# run from fast_main
+
+# Adds vertical climbing abilities when wearing crampons and not holding anything
+execute at @s[gamemode=!spectator,tag=gm4_mountaineering_using_crampons,predicate=gm4_mountaineering:hands_empty] anchored eyes if block ^ ^ ^1 #gm4_mountaineering:climbable_blocks run function gm4_mountaineering:climbing_effects/add_climb_effects
+
+# Give resistance before landing to reduce fall damage when holding poles in both hands
+execute at @s[gamemode=!spectator,predicate=gm4_mountaineering:holding_poles] if block ~ ~-2 ~ minecraft:air run effect give @s resistance 1 1
```
Since these both use `at @s` you should just include the `at @s` in `player_fast_main` and use `if entity` here

BluePsychoRanger commented on contributors.json:
```diff
@@ -114,5 +114,8 @@
         "links": ["https://twitter.com/wumpacraft"]
+    },
+    {
+        "name":"foodiebonus"
```
```suggestion
        "name":"foodiebonus",
        "links": []
```
I think the field needs to exist. CC @misode

AdequateUsername commented:
To summarise the changes:
everything _should_ now be in order (I think), with the glaring exception of the shulker-teleporting stuff which I'll try and work out

BluePsychoRanger commented on gm4_mountaineering/data/gm4_mountaineering/functions/climbing_effects/reset_shulker_peek.mcfunction:
```diff
@@ -0,0 +1,7 @@
+# at @s
+# run from add_climb_effects
+
+execute as @e[type=shulker, tag=gm4_mountaineering_climb_assist, tag=gm4_mountaineering_new_shulker] at @s run data merge entity @s {Peek:10}
+tag @e[type=shulker, tag=gm4_mountaineering_climb_assist, tag=gm4_mountaineering_new_shulker] remove gm4_mountaineering_new_shulker
```
```suggestion
execute as @e[type=shulker, tag=gm4_mountaineering_climb_assist, tag=gm4_mountaineering_new_shulker] run data merge entity @s {Peek:10}
tag @e[type=shulker] remove gm4_mountaineering_new_shulker
```
minor optimizations: it's actually better not to check tags when removing them. also you don't need `at @s` when just setting the data

Bloo-dev reviewed (approved):
We will do an internal in-game test of this now, therefore I'm giving the ok to run the actions on this.

TheThanathor commented:
We're going to look into this more later, but for now here's a few things about the Crampons you can work on:
- shulkers seem to detect snow layers as blocks
- shulkers can spawn in air (seems to be related to being 2 blocks above the snow layer?)
- crampons don't give armor

AdequateUsername commented:
Thanks for pointing those out, I'll get a look at it at some point (_hopefully_) within a week, as soon as I can.

EpyonProjects commented:
I personally would like to see the skis speed increase persist as long as you're consistently running on snow (there shouldn't be that much of a speed decrease if you're coasting at a speed, if you're still running/skiing on snow, though I'm not a skier so I don't know how true that is) and if possible, reacquire the increased speed after landing if the player decides to jump with the increased snow speed.

A strange request I have is that I believe skis should have a default movement speed debuff (they're difficult to walk in! regardless of snow or not), and then quickly increment speed.  would be more speed increase intervals.

tl;dr
Remove the "remove speed if no longer traveling downhill" part of Skis.  Default slow movement speed, add more speed increments but also activate speed increments quicker, and potentially slight max speed increase

AdequateUsername commented:
Thanks for the comments! Some tweaks made in response to both:
- shulkers no longer spawn in air (I'm not sure what is meant about snow layers though? I couldn't reproduce any issues to do with them)
- Shulkers won't spawn on top of each other (not an issue raised here, but discovered while fixing the air problem)
- Crampons will give correct armour now
- Speed will persist for much longer on snow (not forever, but somewhere around 10 seconds on flat), and will be easier to maintain. The same vertical velocity thresholds for getting the speed have been kept, but the threshold for keeping the speed effects have been lowered (players should only need to drop by 1 block to keep it). The speed increase is also greater.
- To balance the longer speed time, if a player goes uphill, the additional speed effects are lost faster

I'm not sure about the gradual speed increases when starting out. It does take longer to get up to speed than, say, walking normally, but I don't think the effect is significant enough that implementing it wouldn't feel clunky to use. If more people echo the want for this feature, I'm happy to put it in, but I think it would feel somewhat strange to use, given that gradual speed increases are uncommon in the rest of the game (sprinting, minecarts, even boats to a degree).
But what I would kind of like is creating the 'ice effect' so the player slides around a bit more - I don't know if this is actually possible though, and don't see a nice (certainly not obvious) way to implement it.

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/climbing_effects/add_climb_effects.mcfunction:
The climbing mechanics still feel kind of awkward, for example, when you jump up from a 5 high wall you cannot actually reach the top. I'd suggest playing around with this a little bit to try and make it feel nicer.

AdequateUsername replied:
I agree the odd-height walls are hard to climb sometimes. This should be a little easier now.

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/climbing_effects/damage_shulkers.mcfunction:
```diff
@@ -0,0 +1,8 @@
+# increase the score of each climbing shulker, and kill after reaching 6
+# @s = shulker with tag gm4_mountaineering_climb_assist
+# at @s
+# run from player_main
```
```suggestion
# run from main
```
This will remove shulkers the player is standing on, not sure if that's intended

AdequateUsername replied:
That behaviour is intended, since I don't think it should be possible to hang on to a wall forever; eventually, you get tired (or I do). But if it is preferred by more reviewers that a player can stand on a wall forever, that can be changed.

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/init.mcfunction:
```diff
@@ -0,0 +1,22 @@
+scoreboard objectives add gm4_mountaineering_jump_counter minecraft.custom:minecraft.jump
+scoreboard objectives add gm4_mountaineering_climbing_shulker_countdown dummy
+
+schedule function gm4_mountaineering:main 10t
+schedule function gm4_mountaineering:fast_main 10t
```
there's not really a need to schedule these 10t

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/load.mcfunction:
this file can be deleted, it'll be created by beet when the module is build

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/main.mcfunction:
```diff
@@ -0,0 +1,15 @@
+scoreboard players reset @a gm4_mountaineering_jump_counter
+
+# Apply damage to any climbing shulkers
+execute as @e[type=minecraft:shulker,tag=gm4_mountaineering_climb_assist] at @s run function gm4_mountaineering:climbing_effects/damage_shulkers
```
While this method works, it might be better to kill shulkers when there are no players with the `gm4_mountaineering_using_crampons` tag are standing on it. This is just a suggestion of how it might be improved though, you'll have to test some methods out yourself

TheThanathor commented on gm4_mountaineering/data/load/tags/functions/load.json:
the entire load folder can be removed, beet will build that

TheThanathor commented on gm4_mountaineering/data/load/tags/functions/gm4_mountaineering.json:
the entire load folder can be removed, beet will build that

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/predicates/on_snowy_block_or_air.json:
```diff
@@ -0,0 +1,22 @@
+{
+  "condition": "minecraft:alternative",
```
```suggestion
  "condition": "minecraft:any_of",
```

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/predicates/on_snowy_block.json:
```diff
@@ -0,0 +1,32 @@
+{
+  "condition": "minecraft:alternative",
```
```suggestion
  "condition": "minecraft:any_of",
```

TheThanathor reviewed (changes requested):
Pretty much everything looks good, but there's some updates needed to make the module compatible with 1.20+, I'd also suggest trying to improve climbing a bit more, but that is up to you.

AdequateUsername commented:
Thanks for the comments. Translations have been updated to the latest standard, and the formatting changed as requested.

Some notes about the climbing:
- I have now made a small change that (should) make it much easier to climb a height-5 wall (or any odd-height wall).
- I have kept the countdown-based shulker-killing mechanism; as I don't think players should be able to hold onto a wall forever (it's unrealistic for someone to do this).
- I think much of the remaining clunkiness will be coming from the shulker-spawn function not running during a jump (since the shulker-spawn function is running every 4t, but a jump may be finished within that). I don't think there's a way to improve that without increasing the number of times the function runs, which I'm not keen to do.

I'll continue to think of possible improvements to the climbing system over the next days, but this is where I am now.

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/init.mcfunction:
```diff
@@ -0,0 +1,19 @@
+scoreboard objectives add gm4_mountaineering_poles_jump_boost dummy
+scoreboard objectives add gm4_mountaineering_jump_counter minecraft.custom:minecraft.jump
+scoreboard objectives add gm4_mountaineering_climbing_shulker_countdown dummy
+
```
You do actually still need these here, my comment was that they don't need to be 10t when 1t works fine
```suggestion

schedule function gm4_mountaineering:main 1t
schedule function gm4_mountaineering:fast_main 1t
```

TheThanathor reviewed (changes requested):
two minor problems, other than that it looks good!

TheThanathor commented on gm4_mountaineering/data/gm4_mountaineering/functions/climbing_effects/add_climb_effects.mcfunction:
```diff
@@ -0,0 +1,19 @@
+execute if entity @s[y_rotation=45..135] positioned ~-1 ~-1.5 ~ align xyz unless entity @e[type=shulker, tag=gm4_mountaineering_climb_assist, dx=0.5, dz=0.5, dy=1] unless block ~0.5 ~1 ~0.5 minecraft:air run summon minecraft:shulker ~0.5 ~1000 ~0.5 {Color:7,AttachFace:4,Invulnerable:1b,NoAI:1b,NoGravity:1b,PersistenceRequired:1b,Silent:1b,Tags:[gm4_mountaineering_climb_assist, gm4_mountaineering_ ...
+
+# teleport shulker down 999 blocks to bring them into position (999 to put the shulker one block below feet)
+execute positioned ~ ~1000 ~ as @e[type=shulker, tag=gm4_mountaineering_climb_assist, tag=gm4_mountaineering_new_shulker] at @s run teleport @s ~ ~-999 ~
```
in multiplayer this function runs twice, meaning the first shulker summoned gets teleported twice, you'll need to remove the tag immediately afterwards and use a different tag to track the shulker peeking

TheThanathor reviewed (changes requested):
couple more points found during the meeting

BluePsychoRanger commented:
The Ski Poles should be called "Ski Pole" which would make it more clear that the player needs two of them and that they need one in each hand. We also discussed making them unstackable, but that can be done later and isn't an immediate issue. Ping me on discord for info about making items unstackable.

Bloo-dev commented:
We now have a planned timeline for this:

Once all the multiplayer issues (second review by @TheThanathor) and the fixes listed by @TheThanathor above have been applied, this will get one more feature-check just to make sure all the item names and Advancements follow Gamemode 4 Guidelines and be merged after.
However, the module will get a note of _Experimental Module_, similar to [Combat Expanded](https://gm4.co/modules/combat-expanded).

This is due to us having ideas on how to make the skis feel better and us wanting to apply some advanced beet stuff to make the Crampons easier to use:
- The skies may be better of being a sled, similar to the sleds found in Origin Realms (but this would look very bad without the resource pack
- The ski poles should be unstackable
- Using beet and display entities the shulkers should also summon a display entity which mimics the block being climbed on, but pushed out by one pixel (creating the impression of a ledge the player stands on)

These changes are not to be made by you (unless you really want to) but will be added as a later update.

AdequateUsername commented on gm4_mountaineering/data/gm4_mountaineering/functions/player_fast_main.mcfunction:
```diff
@@ -0,0 +1,11 @@
+
+# Adds vertical climbing abilities when wearing crampons and not holding anything
+execute if entity @s[gamemode=!spectator,tag=gm4_mountaineering_using_crampons,predicate=gm4_mountaineering:hands_empty] anchored eyes if block ^ ^ ^1 #gm4_mountaineering:climbable_blocks run function gm4_mountaineering:climbing_effects/add_climb_effects
+execute if entity @s[gamemode=!spectator,tag=gm4_mountaineering_using_crampons,predicate=gm4_mountaineering:hands_empty] anchored eyes if block ^ ^-1 ^1 #gm4_mountaineering:climbable_blocks run function gm4_mountaineering:climbing_effects/add_climb_effects
```
I've added `unless block ^ ^ ^1 #gm4_mountaineering:climbable_blocks` into the second condition - that is, only run `add_climb_effects` in the second check if the condition for the first check was not met. I think this should achieve the same result? (in my own test, it makes climbing a little cleaner)

TheThanathor replied:
That will work as well, but using the score method is a little bit more optimized as it only does the score check if the first check succeeds, by using `unless block` it needs to do a block check, which is worse for performance

AdequateUsername replied:
Alright, makes sense. Changed to a scoreboard approach (it seems a little clunky, so I don't know if there's a better way to do it, but I know at least that it works)

TheThanathor reviewed (approved):
looks good to me! There's def some opportunity to improve things with the new ride command, but that can be a thing for a later update

AdequateUsername commented:
Final tweaks:
- scoreboard-based approach to prevent add_climb_effects running twice when the two possible conditions are both met
- the ski pole can no longer be stacked

AdequateUsername commented on gm4_mountaineering/data/gm4_mountaineering/functions/check_recipes.mcfunction:
```diff
@@ -0,0 +1,5 @@
+# Checks for recipes then runs the relevant function to create mountaineering gear
+
+execute if score $crafted gm4_crafting matches 0 store success score $crafted gm4_crafting if score $slot_count gm4_crafting matches 7 if score $stack_size gm4_crafting matches ..1 if data storage gm4_custom_crafters:temp/crafter {Items:[{Slot:0b,id:"minecraft:iron_ingot"},{Slot:2b,id:"minecraft:iron_ingot"},{Slot:3b,id:"minecraft:iron_ingot"},{Slot:4b,id:"minecraft:iron_boots"},{Slot:5b,id:"mine ...
+execute if score $crafted gm4_crafting matches 0 store success score $crafted gm4_crafting if score $slot_count gm4_crafting matches 3 if score $stack_size gm4_crafting matches ..64 if data storage gm4_custom_crafters:temp/crafter {Items:[{Slot:0b,id:"minecraft:tripwire_hook"},{Slot:3b,id:"minecraft:stick"},{Slot:6b,id:"minecraft:stick"}]} run loot replace block ~ ~ ~ container.0 loot gm4_mountai ...
```
Hups, I missed that. Changed it.

SpecialBuilder32 commented:
Alright it looks like this is all set to merge :tada: Thanks for all your patience and hardwork @AdequateUsername. 

We've got a marketing plan laid out for the next few weeks to publicize newish modules in various discord servers: Mountaineering is slated for next week's spotlight.
